pippy/PipelineStage.py

```diff
@@ -709,16 +814,28 @@ def forward(self, *args, **kwargs):
 
         output_chunks = [None] * self.chunks
 
-        # Forward pass of all chunks
-        for chunk in range(self.chunks):
-            s = self.streams[chunk % self.nstreams]
-            with torch.cuda.stream(s):
-                output, send_reqs = self.forward_one_chunk(
-                    chunk, args_split, kwargs_split, fwd_cache
-                )
-                all_send_reqs += send_reqs
-                # Prepare for final output merge or reduction
-                output_chunks[chunk] = output
+        if self.inner_depth > 1:
+            # Forward pass of all chunks
+            for chunk in range(self.chunks):
+                s = self.streams[chunk % self.nstreams]
+                with torch.cuda.stream(s):
+                    output, send_reqs = self.forward_one_chunk_ipipe(
+                        chunk, args_split, kwargs_split, fwd_cache
+                    )
+                    all_send_reqs += send_reqs
+                    # Prepare for final output merge or reduction
+                    output_chunks[chunk] = output
+        else:
+            # Forward pass of all chunks
+            for chunk in range(self.chunks):
+                s = self.streams[chunk % self.nstreams]
+                with torch.cuda.stream(s):
+                    output, send_reqs = self.forward_one_chunk(
+                        chunk, args_split, kwargs_split, fwd_cache
+                    )
+                    all_send_reqs += send_reqs
+                    # Prepare for final output merge or reduction
+                    output_chunks[chunk] = output
```

Comment:
The code has lots of duplication here, can you kindly consolidate a little bit?

Reply:
cleaned in the subsequent commit